File: dotnet/GHIElectronics.DUELink/FileSystem.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net;
using System.Text;
using System.Threading.Tasks;
using static System.Runtime.InteropServices.JavaScript.JSType;

namespace GHIElectronics.DUELink {
    public partial class DUELinkController {
        public class FSController {
            SerialInterface serialPort = null;
            StreamController stream;

            public FSController(SerialInterface port, StreamController stream) {
                this.serialPort = port;
                this.stream = stream;   
            } 
           

            int ParseReturn() {
                var ret = this.serialPort.ReadResponse();

                if (ret.success) {
                    try {
                        var value = int.Parse(ret.response);

                        return value;
                    }
                    catch {

                    }
                }
                return -1;
            }
            public int Mount(int type, int cs, int baud, int max_handle) {

                var cmd = string.Format("FsMnt({0},{1}, {2}, {3})", type, cs, baud, max_handle);

                this.serialPort.WriteCommand(cmd);

                return this.ParseReturn();
            }

            public int UnMount() {
                this.serialPort.WriteCommand("FsUnMnt()");

                return this.ParseReturn();
            }

            public int Format(int type, int cs, int baud) {

                var cmd = string.Format("FsFmt({0},{1}, {2})", type, cs, baud);

                this.serialPort.WriteCommand(cmd);

                return this.ParseReturn();
            }

            public int Open(string path, int mode) {

                var cmd = string.Format("FsOpen(\"{0}\",{1})", path, mode);

                this.serialPort.WriteCommand(cmd);

                return this.ParseReturn();
            }

            public int Close(int handle) {

                var cmd = string.Format("FsClose({0})", handle);

                this.serialPort.WriteCommand(cmd);

                return this.ParseReturn();
            }

            public int Write(int handle, byte[] data) {

                //var write_array = string.Empty;

                //write_array = "[";

                //for (var i = 0; i < data.Length; i++) {
                //    write_array += data[i];

                //    if (i < data.Length - 1)
                //        write_array += ",";
                //}

                //write_array += "]";

                //var cmd = string.Format("FsWrite({0}, {1})", handle, write_array);

                //this.serialPort.WriteCommand(cmd);

                //return this.ParseReturn();

                // declare b9 to write
                var cmd = $"dim b9[{data.Length}]";
                this.serialPort.WriteCommand(cmd);
                this.serialPort.ReadResponse();

                var ret = this.stream.WriteBytes("b9", data);

                cmd = string.Format("FsWrite({0}, b9)", handle);
                this.serialPort.WriteCommand(cmd);
                return this.ParseReturn();
            }

            public int Read(int handle, byte[] data) {
                var count = data.Length;                

                var cmd = string.Format("dim b9[{0}]", data.Length);
                this.serialPort.WriteCommand(cmd);
                this.serialPort.ReadResponse();

                cmd = string.Format("FsRead({0}, b9)", handle);
                this.serialPort.WriteCommand(cmd);
                this.serialPort.ReadResponse();

                // use stream to read b9
                var ret = this.stream.ReadBytes("b9", data);

                return ret;
            }

            public int Sync(int handle) {

                var cmd = string.Format("FsSync({0})", handle);

                this.serialPort.WriteCommand(cmd);

                return this.ParseReturn();
            }

            public int Seek(int handle, int offset) {

                var cmd = string.Format("FsSeek({0},{1})", handle, offset);

                this.serialPort.WriteCommand(cmd);

                return this.ParseReturn();
            }

            public int Tell(int handle) {

                var cmd = string.Format("FsTell({0})", handle);

                this.serialPort.WriteCommand(cmd);

                return this.ParseReturn();
            }

            public int Delete(string path) {

                var cmd = string.Format("FsDel(\"{0}\")", path);

                this.serialPort.WriteCommand(cmd);

                return this.ParseReturn();
            }

            public int Find(string path) {

                var cmd = string.Format("FsFind(\"{0}\")", path);

                this.serialPort.WriteCommand(cmd);

                return this.ParseReturn();
            }

            public int Size(string path) {

                var cmd = string.Format("fsfsz(\"{0}\")", path);

                this.serialPort.WriteCommand(cmd);

                return this.ParseReturn();
            }

        }
    }
}

File: micropython/test/duelink/filesystem.py

class FileSystemController:    

    def __init__(self, serialPort,stream):
        self.serialPort = serialPort
        self.stream = stream

    def __ParseReturn(self)->int:
        r,s = self.serialPort.ReadResponse()
        
        if (r):
            try:
                value = int(s)
                return value
            except:
                return -1
        return -1

    def Write(self, type: int, cs: int, baud: int, max_handle: int)->int:            
        cmd = f"FsMnt({type}, {cs}, {baud}, {max_handle})"
        self.serialPort.WriteCommand(cmd)
        
        return self.__ParseReturn()

    def UnMount(self)->int:        
        self.serialPort.WriteCommand("FsUnMnt()")
        return self.__ParseReturn()
    
    def Format(self, type: int, cs: int, baud: int)->int:            
        cmd = f"FsFmt({type}, {cs}, {baud})"
        self.serialPort.WriteCommand(cmd)
        
        return self.__ParseReturn()
    
    def Open(self, path: str, mode: int)->int:            
        cmd = f"FsOpen({path}, {mode})"
        self.serialPort.WriteCommand(cmd)
        
        return self.__ParseReturn()
    
    def Close(self, handle: int)->int:            
        cmd = f"FsClose({handle})"
        self.serialPort.WriteCommand(cmd)
        
        return self.__ParseReturn()
    
    def Write(self, handle: int, data: bytes)->int:  
        count = len(data)          
        
        cmd = f"dim b9[{count}]"
        self.serialPort.WriteCommand(cmd)
        self.serialPort.ReadResponse()

        ret = self.stream.WriteBytes("b9", data)

        cmd = f"FsWrite({handle}, b9)"
        self.serialPort.WriteCommand(cmd)
        return self.__ParseReturn()
    
    def Read(self, handle: int, data: bytearray)->int:  
        count = len(data)          
        
        cmd = f"dim b9[{count}]"
        self.serialPort.WriteCommand(cmd)
        self.serialPort.ReadResponse()
        
        cmd = f"FsRead({handle}, b9)"
        self.serialPort.WriteCommand(cmd)    
        self.serialPort.ReadResponse()

        ret = self.stream.ReadBytes("b9",data)
        return ret
    
    def Sync(self, handle: int)->int:            
        cmd = f"FsSync({handle})"
        self.serialPort.WriteCommand(cmd)
        
        return self.__ParseReturn()
    
    def Seek(self, handle: int, offset: int)->int:            
        cmd = f"FsSeek({handle},{offset})"
        self.serialPort.WriteCommand(cmd)
        
        return self.__ParseReturn()
    
    def Tell(self, handle: int)->int:            
        cmd = f"FsTell({handle})"
        self.serialPort.WriteCommand(cmd)
        
        return self.__ParseReturn()
    
    def Delete(self, path: str)->int:            
        cmd = f"FsDel({path})"
        self.serialPort.WriteCommand(cmd)
        
        return self.__ParseReturn()
    
    def Find(self, path: str)->int:            
        cmd = f"FsFind({path})"
        self.serialPort.WriteCommand(cmd)
        
        return self.__ParseReturn()
    
    def Size(self, path: str)->int:            
        cmd = f"Fsfsz({path})"
        self.serialPort.WriteCommand(cmd)
        
        return self.__ParseReturn() 
